README.md
# Master of Ceremonies

Interfaces for making MATSim configuration less painful.

Modify, generate, diff and debug configs from the command line.

## Contents

* [Installation](#installation)
* [Usage](#usage)
* [Command Line](#command-line)
* [Programming Interface](#programming-interface)
* [Find](#find)
* [Validation](#validation)
* [Updating MC for config changes](#updating-mc-for-config-changes)
* [Tests](#tests)
* [Contact](#contacts)
* [Todo](#todo)

## Installation

MC requires python3.5 or above.

```{bash}
git@github.com:arup-group/mc.git
cd mc
virtualenv -p python3.7 venv
source venv/bin/activate
pip3 install -e .
pytest
```

Although we recomment using a virtual environment as shown above (`virtualenv...` & `source...`), MC requirements are minimal so you may get away with your default environment.

## Usage

MC has a Command Line Interface (CLI), but also a useful pythonic API for bespoke applications.

The CLI provides a number of features, introduced below. These include some general purpose functionality such as `diff`, as well as some bespoke features required for other tooling, such as `paramreplace`, `step` and `autostep`.

We are also interested in maintaining a `debug` command to catch common config mistakes. If you have a feature in mind, such as debugging of a common config mistake please add it to the project issues.

## Command Line

The CLI commands are pretty explorable via help, start with `mc --help`:

```{bash}
Usage: mc [OPTIONS] COMMAND [ARGS]...

  Command line interface for MC.

Options:
  --help  Show this message and exit.

Commands:
  autostep      Read config, apply overrides and write out.
  build         Build a config with defined sub-pops, modes & activities.
  convert       Read an existing config and write as xml or json.
  debug         Debug a config.
  diff          Simple diff two configs.
  fill          Read a wildcarded config, apply overrides and write.
  find          Find and print config components to terminal.
  gen           Generate a template config: empty|default|test.
  matchreplace  Read wildcarded config, apply overrides and write.
  paramreplace  Read config, apply overrides and write out.
  print         Print a config to terminal.
  step          Read config, apply overrides and write out.
```

### Print Example

(from MC root)

```{bash}
❯ mc print tests/test_data/test_config.xml
```

```{bash}
module {'name': 'global'}
   param {'name': 'coordinateSystem', 'value': 'EPSG:27700'}
   param {'name': 'insistingOnDeprecatedConfigVersion', 'value': 'true'}
   param {'name': 'numberOfThreads', 'value': '32'}
   param {'name': 'randomSeed', 'value': '4711'}
 module {'name': 'network'}
   param {'name': 'inputCRS', 'value': 'null'}
   param {'name': 'inputChangeEventsFile', 'value': 'null'}
   param {'name': 'inputNetworkFile', 'value': '~/test/network.xml'}
   param {'name': 'laneDefinitionsFile', 'value': 'null'}
   param {'name': 'timeVariantNetwork', 'value': 'false'}
   ...
```

### Diff Example

(from MC root)

```{bash}
❯ mc diff tests/test_data/test_config.xml tests/test_data/test_config_v12.xml
```

```{bash}
- param@plans: insistingOnUsingDeprecatedPersonAttributeFile
- param@plans: inputPersonAttributesFile
+/- param@strategy: fractionOfIterationsToDisableInnovation: 0.95 -> 0.9
```

### Find Example

(from MC root)

```{bash}
❯ mc find tests/test_data/test_config.xml coordinateSystem
```

```{bash}
param {'name': 'coordinateSystem', 'value': 'EPSG:27700'}
```

You can read about more complex addressing in [Find](##find).

## Programming Interface

```{python}
from mc import build
```

To facilitate a beautiful future where humans never need edit xml and machines start to explore MATSim simulation config parameter spaces themselves, MC makes a dictionary-like `mc.build.Config` object available. `Config` objects can read and write to MATSim `.xml` config format (and `.json` just in case).

```{python}
config = build.Config(path='tests/test_data/test_config.xml')
config.write(path='temp.json')
config2 = build.Config(path='temp.json')
config == config2
```

```{python}
True
```

### Quick Reminder

MATSim configs consist of modules, these modules can contain either parameters and/or parameter***sets*** (ie sets of paramaters). Paramater***sets*** can then contain either parameters and/or more parameter***sets***.

### Setting and Getting

`Config` objects consist of nested `Modules`, `ParamSets` and `Params` (just like a MATSim xml formatted config). All of which will behave like a nested set of dicts. Icluding supporting getting and setting methods:

```{python}
# get and print module contents:
config['plans'].print()
```

```{python}
module {'name': 'plans'}
    param {'name': 'inputPlansFile', 'value': 'test_inputs/population.xml'}
    param {'name': 'inputPersonAttributesFile', 'value': 'test_inputs/attributes.xml'}
    param {'name': 'subpopulationAttributeName', 'value': 'subpopulation'}
```

```{python}
 # set and print a single param:
config['plans']['inputPlansFile'] = 'test_inputs/new_population.xml'
print(config['plans']['inputPlansFile'])
```

```{python}
test_inputs/new_population.xml
```

### Nested Setting and Getting

Nested setting is allowed, for example for an empty `Config` a new module, paramset and param
 can be set together:

```{python}
empty_config = build.Config()
empty_config['global']['coordinateSystem'] = 'EPSG:27700'
...
empty_config.print()
```

```{python}
module {'name': 'global'}
    param {'name': 'coordinateSystem', 'value': 'EPSG:27700'}
```

### Validation of Keys and Values

Setting requires that keys are valid:

```{python}
empty_config = build.Config()
empty_config['NotAModule']['coordinateSystem'] = 'EPSG:27700'
```

```{python}
...
KeyError: "key:'NotAModule' not found in modules"
```

...and that values are valid:

```{python}
empty_config['global']['coordinateSystem'] = {"crs": "27700"}

```

```{python}
INFO creating new empty module: global
ValueError: Please use value of either type ParamSet, Param or str
```

Nested objects can be explicitly accessed via the parent attributes, ie `.modules.values()`, `.parametersets.values()`, `params.values()`. Valid keys (ignoring suffixes) for nestable
objects via `.valid_keys`. Note that valid keys and values are hardcoded in `mc.valid`.

### Unique Parameterset Keys

MATSim configurations include parameter***sets*** for which the unique identification (such as the mode or subpopulation) is contained as a parameter. So that we can provide unique keys for the parameter***set***, we therefore suffix parameter***set*** keys as `<paramset_name>:<uid>` where `uid` is the appropriate parameterset
subpopulation, mode or activity:

```{python}
empty_config.write(path=Path('temp.xml'))
empty_config['planCalcScore']['scoringParameters:high_income']['modeParams:car']['monetaryDistanceRate'] = '-0.0001'
```

```{python}
INFO creating new empty module: planCalcScore
INFO creating new empty parameterset: scoringParameters:high_income
INFO creating new empty parameterset: modeParams:car
```

In this case we have created a scoring parameterset for the `high_income` subpopulation and a mode parameterset for `car`.

When writing to `.xml` these suffixes are ignored, writing a clean usable config. When reading in
an existing config, these suffixes are automatically generated.

**WARNING: Not specifying a suffix will generally be assumed as using `<paramset_name>:default`.**

## Find

Both the CLI and API support searches for config components using an addressing system. The addressing system uses a string format with `/` to denote a nested parameterset or parameter. This allows changes to be easilly passed as commands via common web interfaces for example.

```{python}
config = build.Config(path='tests/test_data/test_config.xml')
search = config.find("plans/inputPlansFile")
for i in search:
  i.print()
```

```{python}
param {'name': 'inputPlansFile', 'value': '~/test/population.xml.gz'}
```

Find is returning a list because it supports partial addresses which result in multiple finds:

```{python}
search = config.find("modeParams:car/monetaryDistanceRate")
for i in search:
  i.print()
```

```{python}
param {'name': 'monetaryDistanceRate', 'value': '-0.0001'}  # eg subpopulation A
param {'name': 'monetaryDistanceRate', 'value': '-0.0001'}  # eg subpopulation B
```

Addresses can omit components, for example if we want to look at all `monetaryDistanceRates` for the default subpopulation (ie from `scoringParameters:default`):

```{python}
search = config.find("scoringParameters:default/monetaryDistanceRate")
# this is equivalent to "*/scoringParameters:default/*/monetaryDistanceRate"
for i in search:
  i.print()
```

```{python}
param {'name': 'monetaryDistanceRate', 'value': '-0.0'}  # eg walk
param {'name': 'monetaryDistanceRate', 'value': '-0.0'}  # eg bike
param {'name': 'monetaryDistanceRate', 'value': '-0.001'}  # eg pt
param {'name': 'monetaryDistanceRate', 'value': '-0.0001'}  # eg car
```

Or more simply we can get all `monetaryDistanceRates`:

```{python}
search = config.find("monetaryDistanceRate")
# this is equivalent to "*/*/*/monetaryDistanceRate"
# this is equivalnet to "*/scoringParameters:*/modeParams:*/monetaryDistanceRate"
for i in search:
  i.print()
```

```{python}
param {'name': 'monetaryDistanceRate', 'value': '-0.0'}  # eg subpop A walk
param {'name': 'monetaryDistanceRate', 'value': '-0.0'}  # eg subpop A bike
param {'name': 'monetaryDistanceRate', 'value': '-0.001'}  # subpop A eg pt
param {'name': 'monetaryDistanceRate', 'value': '-0.0001'}  # subpop A eg car
param {'name': 'monetaryDistanceRate', 'value': '-0.0'}  # eg subpop B walk
param {'name': 'monetaryDistanceRate', 'value': '-0.0'}  # eg subpop B bike
param {'name': 'monetaryDistanceRate', 'value': '-0.001'}  # eg subpop B pt
param {'name': 'monetaryDistanceRate', 'value': '-0.0001'}  # eg subpop B car
```

In the examples above, you can see that wildcarding with `*` can be used to return 'all' config elements. The `*` operator tells the find method to search all at a given level. As shown above, it is useful for returning all elements within a parameterset or explicitly describing levels to search.

### Find and Set

Note that the `find` method is returning a reference to the found config objects, which can then be set:

```{python}
config.find("plans/inputPlansFile")[0] = "NEW/PATH"
```

## Validation

MC has a built-in representation of a valid config structure, specifically the viable names of modules, parametersets and parameters. When reading in an existing config or adding new components, MC will throw validation errors if the valid config structure is not maintained.

```{python}
empty_config = build.Config()
empty_config['NotAModule']['coordinateSystem'] = 'EPSG:27700'
```

```{python}
...
KeyError: "key:'NotAModule' not found in modules"
```

This system is useful for preventing typos, but has to be
[maintained and updated for changes to valid configs](#updating-mc-for-config-changes). The valid mapping is described in the `mc.valid` module.

## Updating MC for Config Changes

An example of how to update the validation mapping can be seen with the addition of a new `hermes` module:

```{xml}
  <module name="hermes" >
      <param name="endTime" value="32:00:00" />
      <param name="flowCapacityFactor" value="0.01" />
      <param name="mainMode" value="car" />
      ...
  </module>
```

In order to make this hermes module available to MC's validation, the following is added to `mc/valid.py`:

```{json}
  "hermes": {
      "params": {
            "mainMode": "car",
            "endTime": "36:00:00",
            "flowCapacityFactor": "0.01",
            ...
      }
  },
```

## Tests

```{bash}
    python -m pytest -vv tests
```

To generate XML & HTML coverage reports to `reports/coverage`:

```{bash}
    ./scripts/code-coverage.sh
```

## Contact

fred.shone@arup.com

## Further Work

* Add more and better debugging
* Update default configurations
* Print a high level summary to terminal (scaling/subpops, modes etc)
* Add detailed value validation, for example acceptable integer ranges
* Add further debugging by comparing config to other inputs such as transit schedules


RELEASE_NOTES.md
0.0.1 - 07/12/2021 - initial release


mc/valid.py
"""
Dictionary of valid config names for modules, paramsets and params.
"""

VALID_MAP = {

    "modules": {

        "swissRailRaptor": {
            "params": {
                "scoringParameters": "Default",
                "transferPenaltyBaseCost": "0.0",
                "transferPenaltyCostPerTravelTimeHour": "0.0",
                "transferPenaltyMaxCost": "Infinity",
                "transferPenaltyMinCost": "-Infinity",
                "useIntermodalAccessEgress": "false",
                "useModeMappingForPassengers": "false",
                "useRangeQuery": "false",
            },
            "parametersets": {
                "intermodalAccessEgress": {
                    "params": {
                        "mode": "walk",
                        "radius": "1000"
                    }
                },
                "modeMapping": {
                    "params": {
                        "routeMode": "bus",
                        "passengerMode": "bus"
                    }
                }
            }
        },

        "SBBPt": {
            "params": {
                "createLinkEventsInterval": "1",
                "deterministicServiceModes": "train",
            }
        },

        "global": {
            "params": {
                "coordinateSystem": "EPSG:27700",
                "numberOfThreads": "32",
                "randomSeed": "4711",
                "insistingOnDeprecatedConfigVersion": "true",
            }
        },

        "network": {
            "params": {
                "inputCRS": "null",
                "inputChangeEventsFile": "null",
                "inputNetworkFile": "./output_network.xml.gz",
                "laneDefinitionsFile": "null",
                "timeVariantNetwork": "false"
            }
        },

        "plans": {
            "params": {
                "inputPlansFile": "./output_plans.xml.gz",
                "inputPersonAttributesFile": "./output_personAttributes.xml.gz",
                "subpopulationAttributeName": "subpopulation",
                "activityDurationInterpretation": "tryEndTimeThenDuration",
                "inputCRS": "null",
                "networkRouteType": "LinkNetworkRoute",
                "removingUnnecessaryPlanAttributes": "false",
            }
        },

        "transit": {
            "params": {
                "insistingOnUsingDeprecatedPersonAttributeFile": "true",
                "useTransit": "true",
                "transitScheduleFile": "./output_transitSchedule.xml.gz",
                "vehiclesFile": "./output_transitVehicles.xml",
                "inputScheduleCRS": "null",
                "transitLinesAttributesFile": "null",
                "transitModes": "pt",
                "transitStopsAttributesFile": "null",
            }
        },

        "multimodal": {
            "params": {
                "createMultiModalNetwork": "true",
                "numberOfThreads": "1",
                "cuttoffValueForNonCarModes": "25.0",
                "dropNonCarRoutes": "true",
                "multiModalSimulationEnabled": "true",
                "simulatedModes": "walk,bike",
                "ensureActivityReachability": "true",
            }
        },

        "travelTimeCalculator": {
            "params": {
                "analyzedModes": "car,bus,walk",
                "filterModes": "true",
                "separateModes": "false",
                "calculateLinkToLinkTravelTimes": "false",
                "calculateLinkTravelTimes": "true",
                "maxTime": "108000",
                "travelTimeAggregator": "optimistic",
                "travelTimeBinSize": "900",
                "travelTimeCalculator": "TravelTimeCalculatorArray",
                "travelTimeGetter": "average"
            }
        },

        "JDEQSim": {
            "params": {
                "carSize": "7.5",
                "endTime": "undefined",
                "flowCapacityFactor": "1.0",
                "gapTravelSpeed": "15.0",
                "minimumInFlowCapacity": "1800.0",
                "squeezeTime": "1800.0",
                "storageCapacityFactor": "1.0"
            }
        },

        "TimeAllocationMutator": {
            "params": {
                "mutationAffectsDuration": "true",
                "mutationRange": "1000.0",
                "useIndividualSettingsForSubpopulations": "false"
            }
        },

        "changeMode": {
            "params": {
                "ignoreCarAvailability": "true",
                "modes": "car,pt"
            }
        },

        "controler": {
            "params": {
                "createGraphs": "true",
                "dumpDataAtEnd": "true",
                "enableLinkToLinkRouting": "false",
                "eventsFileFormat": "xml",
                "firstIteration": "0",
                "lastIteration": "100",
                "mobsim": "qsim",
                "outputDirectory": "home/arup/tii/1p_models/3_default_config/model_out",
                "overwriteFiles": "failIfDirectoryExists",
                "routingAlgorithmType": "Dijkstra",
                "runId": "null",
                "snapshotFormat": "",
                "writeEventsInterval": "1",
                "writePlansInterval": "50",
                "writeSnapshotsInterval": "1"
            }
        },

        "counts": {
            "params": {
                "analyzedModes": "car",
                "averageCountsOverIterations": "5",
                "countsScaleFactor": "1.0",
                "distanceFilter": "null",
                "distanceFilterCenterNode": "null",
                "filterModes": "false",
                "inputCRS": "null",
                "inputCountsFile": "null",
                "outputformat": "txt",
                "writeCountsInterval": "10"
            }
        },

        "facilities": {
            "params": {
                "addEmptyActivityOption": "false",
                "assigningLinksToFacilitiesIfMissing": "true",
                "assigningOpeningTime": "false",
                "facilitiesSource": "none",
                "idPrefix": "",
                "inputCRS": "null",
                "inputFacilitiesFile": "null",
                "inputFacilityAttributesFile": "null",
                "oneFacilityPerLink": "true",
                "removingLinksAndCoordinates": "true"
            }
        },

        "households": {
            "params": {
                "inputFile": "null",
                "inputHouseholdAttributesFile": "null"
            }
        },

        "linkStats": {
            "params": {
                "averageLinkStatsOverIterations": "5",
                "writeLinkStatsInterval": "10"
            }
        },

        "parallelEventHandling": {
            "params": {
                "estimatedNumberOfEvents": "null",
                "numberOfThreads": "32",
                "oneThreadPerHandler": "false",
                "synchronizeOnSimSteps": "true"
            }
        },

        "planCalcScore": {
            "params": {
                "BrainExpBeta": "1.0",
                "PathSizeLogitBeta": "1.0",
                "fractionOfIterationsToStartScoreMSA": "null",
                "learningRate": "1.0",
                "usingOldScoringBelowZeroUtilityDuration": "false",
                "writeExperiencedPlans": "false"
            },

            "parametersets": {
                "scoringParameters": {
                    "params": {
                        "earlyDeparture": "-0.0",
                        "lateArrival": "-18.0",
                        "marginalUtilityOfMoney": "0.0",
                        "performing": "6.0",
                        "subpopulation": "default",
                        "utilityOfLineSwitch": "-1.0",
                        "waiting": "-1.0",
                        "waitingPt": "-1.0"
                    },

                    "parametersets": {
                        "activityParams": {
                            "params": {
                                "activityType": "education",
                                "closingTime": "17:00:00",
                                "earliestEndTime": "undefined",
                                "latestStartTime": "undefined",
                                "minimalDuration": "06:00:00",
                                "openingTime": "08:30:00",
                                "priority": "1.0",
                                "scoringThisActivityAtAll": "true",
                                "typicalDuration": "06:00:00",
                                "typicalDurationScoreComputation": "relative"
                            }
                        },

                        "modeParams": {
                            "params": {
                                "constant": "0.0",
                                "marginalUtilityOfDistance_util_m": "0.0",
                                "marginalUtilityOfTraveling_util_hr": "-6.0",
                                "mode": "car",
                                "monetaryDistanceRate": "-0.0"
                            }
                        }
                    }
                }
            }
        },

        "planscalcroute": {
            "params": {
                "networkModes": "car"
            },

            "parametersets": {
                "teleportedModeParameters": {
                    "params": {
                        "beelineDistanceFactor": "1.3",
                        "mode": "bike",
                        "teleportedModeFreespeedFactor": "null",
                        "teleportedModeSpeed": "4.166666666666667"
                    }
                }
            }
        },

        "ptCounts": {
            "params": {
                "countsScaleFactor": "1.0",
                "distanceFilter": "null",
                "distanceFilterCenterNode": "null",
                "inputAlightCountsFile": "null",
                "inputBoardCountsFile": "null",
                "inputOccupancyCountsFile": "null",
                "outputformat": "null",
                "ptCountsInterval": "10"
            }
        },

        "qsim": {
            "params": {
                "creatingVehiclesForAllNetworkModes": "true",
                "endTime": "36:00:00",
                "flowCapacityFactor": "0.01",
                "insertingWaitingVehiclesBeforeDrivingVehicles": "false",
                "isRestrictingSeepage": "true",
                "isSeepModeStorageFree": "false",
                "linkDynamics": "FIFO",
                "linkWidth": "30.0",
                "mainMode": "car",
                "nodeOffset": "0.0",
                "numberOfThreads": "32",
                "removeStuckVehicles": "false",
                "seepMode": "bike",
                "simEndtimeInterpretation": "null",
                "simStarttimeInterpretation": "maxOfStarttimeAndEarliestActivityEnd",
                "snapshotStyle": "equiDist",
                "snapshotperiod": "00:00:00",
                "startTime": "00:00:00",
                "storageCapacityFactor": "0.01",
                "stuckTime": "10.0",
                "timeStepSize": "00:00:01",
                "trafficDynamics": "queue",
                "useLanes": "false",
                "usePersonIdForMissingVehicleId": "true",
                "usingFastCapacityUpdate": "true",
                "usingThreadpool": "true",
                "vehicleBehavior": "teleport",
                "vehiclesSource": "defaultVehicle"
            }
        },

        "hermes": {
            "params": {
                "mainMode": "car",
                "endTime": "36:00:00",
                "flowCapacityFactor": "0.01",
                "storageCapacityFactor": "0.01",
                "stuckTime": "10.0",
                "useDeterministicPt": "true"
            }
        },

        "scenario": {},

        "strategy": {
            "params": {
                "ExternalExeConfigTemplate": "null",
                "ExternalExeTimeOut": "3600",
                "ExternalExeTmpFileRootDir": "null",
                "fractionOfIterationsToDisableInnovation": "0.9",
                "maxAgentPlanMemorySize": "5",
                "planSelectorForRemoval": "WorstPlanSelector"
            },

            "parametersets": {
                "strategysettings": {
                    "params": {
                        "disableAfterIteration": "-1",
                        "executionPath": "null",
                        "strategyName": "SelectExpBeta",
                        "subpopulation": "unknown",
                        "weight": "0.6"
                    }
                }
            }
        },

        "subtourModeChoice": {
            "params": {
                "chainBasedModes": "car,bike",
                "considerCarAvailability": "false",
                "modes": "car,pt,walk,bike"
            }
        },

        "transitRouter": {
            "params": {
                "additionalTransferTime": "0.0",
                "directWalkFactor": "1.0",
                "extensionRadius": "100.0",
                "maxBeelineWalkConnectionDistance": "100.0",
                "searchRadius": "1000.0"
            }
        },

        "vehicles": {
            "params": {
                "vehiclesFile": "null"
            }
        },

        "vspExperimental": {
            "params": {
                "isAbleToOverwritePtInteractionParams": "false",
                "isGeneratingBoardingDeniedEvent": "false",
                "isUsingOpportunityCostOfTimeForLocationChoice": "true",
                "logitScaleParamForPlansRemoval": "1.0",
                "vspDefaultsCheckingLevel": "ignore",
                "writingOutputEvents": "true"
            }
        }
    }
}
